Added Windowing Regularizers for time averaged outputs.

SteffenMath opened the pull request:
## Proposed Changes
*Give a brief overview of your contribution here in a few sentences.*
Based on the observations of Albring et al. in "Challenges in Sensitivity Computations for (D)DES and URANS" we implemented the windowing regularizers proposed in the work of  Krakos et al. "Sensitivity analysis of limit cycle oscillations" and adapted it to the AD framework of SU2. 
We call the feature in the following "windowing". 
Windowing can be applied if one takes the (windowed) time average of a periodic function, i.e. in the case of an instationary flow field that exhibits periodic behaviour. 
 - We changed the functionality of the output-field-prefix TAVG and D_TAVG from a simple time average to a time averaged weighted with a windowing function. For the direct run this is just a post-processing step and changes happen only in signal_processing_toolbox and Coutput. 
- We adapted the seeding for the adjoint run from a the classic time average to a windowed time average. 
-  Tests have shown, that higher order windowing regularizers, tend to converge quickly with respect to the averaged time span. We have implemented a "timeConvergenceMonitor", that acts similarly to the convergence monitoring of the inner convergence. It applies a cauchy convergence criterion to the chosen windowed time average output coefficients. 
- The cauchy convergence criterion works for direct run, adjoint run and shape optimization scripts
- In case of shape optimization, the script remembers the final time iteration of the direct run. If the final time step is smaller than the "original" final time step stated in the configuration file, it adapts the starting time of the adjoint run; and the time to seed the objective function, such that the seeding does not exceed the original time to end the seeding.


## Related Work
*Resolve any issues (bug fix or feature request), note any related PRs, or mention interactions with the work of others, if any.*



## PR Checklist
*Put an X by all that apply. You can fill this out after submitting the PR. If you have any questions, don't hesitate to ask! We want to help. These are a guide for you to know what the reviewers will be looking for in your contribution.*

- [x] I am submitting my contribution to the develop branch.
- [x] My contribution generates no new compiler warnings (try with the '-Wall -Wextra -Wno-unused-parameter -Wno-empty-body' compiler flags).
- [X] My contribution is commented and consistent with SU2 style.
- [x] I have added a test case that demonstrates my contribution, if necessary.
- [x] I have updated appropriate documentation (Tutorials, Docs Page, config_template.cpp) , if necessary.

pcarruscag commented on Common/src/toolboxes/signal_processing_toolbox.cpp:
```diff
@@ -27,11 +27,115 @@
+      case 3: return BumpWindow(CurTimeIdx, endTimeIdx);
+      default:return 1.0;
+    }
+}
```
Here the type of `fctIdx` should be the enumerator, and the "cases" should be actual names not the numbers.
All these functions can be const or static if you use the globally defined pi.

SteffenMath replied:
GetWndWeight now uses the enum WINDOW_FUNCTION.

pcarruscag commented on Common/src/toolboxes/signal_processing_toolbox.cpp:
```diff
@@ -27,11 +27,115 @@
+    }
+}
+su2double WindowingTools::HannWindow(unsigned long i, unsigned long endTimeIdx){
+  if(i==0) return 0; //catch div by zero error
```
There is no risk of division by 0

SteffenMath replied:
I've removed the unnecessary if-clause.

SteffenMath commented:
> Hi @SteffenMath,
> 
> Welcome to SU2 (it seems it is your first PR), it looks like a good generalisation of the moving averages, I leave some small code comments below. As a general point does this averaging rely on fixed timesteps?
> 
> Cheers,
> Pedro

Hi @pcarruscag, 

thanks for your quick reply and the usefull comments. I've pushed and updated version, that includes changes to your suggestions. 
Regarding your general question: The windowing approach is in theory not dependent on a fixed time-stepsize, however the current implementation relies on a fixed stepsize. This can be changed quite easily in the CWindowingTools.cpp file and the corresponding class, should there be demand.
 
Furthermore, I've added some time convegergence monitoring tools that depend on a windowed time averaging.

Best
Steffen

SteffenMath commented on Common/include/option_structure.hpp:
```diff
@@ -835,6 +835,22 @@ static const map<string, ENUM_SGS_MODEL> SGS_Model_Map = CCreateMap<string, ENUM
+/*!
+ * \brief types of window (weight) functions for cost functional
+ */
+enum WND_FUNCTION {
```
The enum is now called WINDOW_FUNCTION

SteffenMath commented on Common/include/toolboxes/signal_processing_toolbox.hpp:
```diff
@@ -30,39 +30,76 @@
+    ~WindowingTools(){}
+
+    /*! \brief Returns the value of a windowing function given by fctIdx at CurTimeIdx with given endTimeIdx (i.e. endTimeIdx=nTimeIter, if one starts  windowing at time t =0.) */
+    su2double GetWndWeight(int fctIdx, unsigned long CurTimeIdx, unsigned long endTimeIdx);
```
I've deleted both constructor and destructor.

SteffenMath commented on Common/include/toolboxes/signal_processing_toolbox.hpp:
```diff
@@ -30,39 +30,76 @@
+    su2double GetWndWeight(int fctIdx, unsigned long CurTimeIdx, unsigned long endTimeIdx);
+
+protected:
+    const su2double PI_NUMBER = 4.0 * atan(1.0);	/*!< \brief Pi number. */
```
I'm using now the constant defined in "option_structure.hpp"

pcarruscag commented on Common/include/config_structure.hpp:
```diff
@@ -9199,15 +9221,55 @@ class CConfig {
+  * \brief Get the number of convergence monitoring fields for time convergence monitoring.
+  * return Number of convergence monitoring fields.
+  */
+  unsigned short GetnWndConv_Field();
```
Thank you for making the changes, it seems all "Get" methods you added can be const (in general all get methods should be), please add the const qualifier to everything you can.

SteffenMath replied:
Fixed it :)

pcarruscag commented on Common/src/config_structure.cpp:
```diff
@@ -1492,6 +1492,17 @@ void CConfig::SetConfig_Options() {
+  addBoolOption("WND_CAUCHY_CRIT",Wnd_Cauchy_Crit, false);
+   /*!\brief WND_CONV_FIELD
+    * \n DESCRIPTION: Output fields  for the Cauchy criterium for the TIME iteration. The criterium is applied to the windowed time average of the chosen funcion. */
+  addStringListOption("WND_CONV_FIELD",nWndConvField, WndConvField);
```
Here on the "user side" I think options should be as descriptive as possible, to that end WND -> WINDOW.
Also @talbring and @rsanfer prefixed all convergence related options with CONV_ I think we should stick to that convention, please move WINDOW to after CONV or to the end of the option.
As a general question, can the windowing feature be used as a convergence criteria for time domain while the normal Cauchy criteria is used for the inner iterations? (trying to understand if we really need these new options or if the existing ones could be re purposed.)

SteffenMath replied:
> As a general question, can the windowing feature be used as a convergence criteria for time domain while the normal Cauchy criteria is used for the inner iterations? (trying to understand if we really need these new options or if the existing ones could be re purposed.)

Yes the windowing feature can be used as a convergence criterion for the time domain independently of the normal Cauchy criterion. You can use both at the same time, just one or none at all. Therefore I think it is useful to define a extra set  options for the windowed time convergence criterion.
In my experience, the windowed-time-cauchy-convergence-criterion needs different parameters than the normal one. For example in a test case, which exhibits only approximately periodic behaviour, one can use the windowed-time-cauchy-convergence-criterion, but usually, the convergence rate is quite slow and it may be usefull to define an EPSILON value, which is much bigger than a reasonable EPSILON for the inner convergence.

pcarruscag commented on Common/src/config_structure.cpp:
```diff
@@ -6301,6 +6318,19 @@ void CConfig::SetOutput(unsigned short val_software, unsigned short val_izone) {
     cout << "Cauchy series min. value: " << Cauchy_Eps << "." << endl;
     cout << "Number of Cauchy elements: " << Cauchy_Elems << "." << endl;
+    if(Cauchy_Elems <1){
+        SU2_MPI::Error(to_string(Cauchy_Elems) + string(" Cauchy elements are no viable input. Please check your configuration file."), CURRENT_FUNCTION);
```
It seems most of the code has 4 space indentation, make that 2 space (on c++ files) to be consistent with the rest of the code.

pcarruscag commented on SU2_CFD/include/drivers/CDiscAdjSinglezoneDriver.hpp:
```diff
@@ -127,4 +127,10 @@ class CDiscAdjSinglezoneDriver : public CSinglezoneDriver {
+   * \brief gets Convergence on physical time scale, (deaktivated in adjoint case)
+   * \param none
+   */
+  virtual bool GetTimeConvergence();
```
Another const :), small typo in "deactivated"

pcarruscag commented on SU2_CFD/src/drivers/CSinglezoneDriver.cpp:
```diff
@@ -280,21 +280,24 @@ bool CSinglezoneDriver::Monitor(unsigned long TimeIter){
     /*--- Check whether the outer time integration has reached the final time ---*/
 
-    FinalTimeReached     = CurTime >= MaxTime;
+    FinalTimeReached     = CurTime >= MaxTime ;
```
stray space before ;

pcarruscag commented on SU2_CFD/src/meson.build:
```diff
@@ -63,7 +63,8 @@ su2_cfd_src += files(['output/CAdjElasticityOutput.cpp',
                       'output/filewriter/CSU2BinaryFileWriter.cpp',
-                      'output/filewriter/CSU2MeshFileWriter.cpp'])
+                      'output/filewriter/CSU2MeshFileWriter.cpp',
+		      'output/tools/CWindowingTools.cpp'])
```
try to align these things neatly please, I noticed that in some of the member variable descriptions too.

pcarruscag commented on SU2_CFD/src/output/tools/CWindowingTools.cpp:
```diff
@@ -0,0 +1,141 @@
+  su2double currTime = static_cast<su2double>(i);
+  su2double endTime = static_cast<su2double>(endTimeIdx);
+  su2double tau = currTime/endTime;
+  return 1.0/0.00702986*(exp(-1/(tau-tau*tau)));
```
this constant warrants a comment, also more const :)

SteffenMath replied:
The constant "0.00702986" equals the integral of the function "(exp(-1/(tau-tau*tau)))" from 0 to 1. We have to divide by it to normalize the window. I add this as a comment.

pcarruscag commented on SU2_CFD/src/output/tools/CWindowingTools.cpp:
```diff
@@ -0,0 +1,141 @@
+        case 1: val    = HannWindowing();         return val;
+        case 2: val    = HannSquaredWindowing();  return val;
+        case 3: val    = BumpWindowing();         return val;
+        default: val   = NoWindowing();           return val;
```
You missed the enumerator here.

SteffenMath replied:
Fixed it :)

pcarruscag commented on SU2_PY/SU2/eval/design.py:
```diff
@@ -122,8 +122,15 @@ def _eval(self,eval_func,*args):
+            konfig['TIME_ITER'] = state.WND_CAUCHY_DATA['TIME_ITER']
+            konfig['ITER_AVERAGE_OBJ'] = state.WND_CAUCHY_DATA['ITER_AVERAGE_OBJ']
+            konfig['UNST_ADJOINT_ITER'] = state.WND_CAUCHY_DATA['UNST_ADJOINT_ITER']
+
```
I am not familiar with the python framework, but I feel this needs a comment.

SteffenMath replied:
If the windowed time convergence criterion is satisfied, the solver may terminate before it has reached the iteration number TIME_ITER. In this case values of TIME_ITER, ITER_AVERAGE_OBJ and UNST_ADJOINT_ITER must be updated appropriately (this happens in /run/direct.py line 120-127). The code section above updates the configuration file for the adjoint run. 
Indeed this is not obvious and I have added comments to clearify this.

pcarruscag commented on SU2_CFD/src/drivers/CDiscAdjSinglezoneDriver.cpp:
```diff
@@ -270,9 +271,11 @@ void CDiscAdjSinglezoneDriver::SetAdj_ObjFunction(){
   if (time_stepping){
     if (TimeIter < IterAvg_Obj){
-      seeding = 1.0/((su2double)IterAvg_Obj);
+      seeding = windowEvaluator.GetWndWeight(config->GetKindWindow(),TimeIter, IterAvg_Obj-1)/ ((su2double)IterAvg_Obj);
```
Don't use "old style" casts (compiler warning) use static_cast or su2double(IterAvg_Obj)

SteffenMath replied:
Fixed it :)

pcarruscag commented on SU2_CFD/src/drivers/CDiscAdjSinglezoneDriver.cpp:
```diff
@@ -539,3 +541,7 @@ void CDiscAdjSinglezoneDriver::SecondaryRecording(){
+
+bool CDiscAdjSinglezoneDriver::GetTimeConvergence() const{
+  return false;
+}
```
This could be inline in the hpp file.

SteffenMath replied:
Sry forgot about this one. Changed it in the last commit :)

pcarruscag commented on SU2_CFD/src/output/COutput.cpp:
```diff
@@ -129,12 +128,29 @@ COutput::COutput(CConfig *config, unsigned short nDim, bool fem_output): femOutp
+
+  for (unsigned short iField = 0; iField < config->GetnWndConv_Field(); iField++){
+    wndConvFields.emplace_back(config->GetWndConv_Field(iField));
+  }
```
General note (not very important in this case) whenever you know how many things you are inserting into the vector you should `reserve` before.

SteffenMath replied:
Okay, I added it nevertheless. Thanks for the feedback!

pcarruscag commented on SU2_CFD/include/drivers/CDiscAdjSinglezoneDriver.hpp:
```diff
@@ -127,4 +127,10 @@ class CDiscAdjSinglezoneDriver : public CSinglezoneDriver {
+   * \brief gets Convergence on physical time scale, (deactivated in adjoint case)
+   * \param none
+   */
+  virtual bool GetTimeConvergence() const;
```
The virtual specifier you should use here is `override` i.e. `bool GetTimeConvergence() const override`, this immediately says to anyone reading the code that you are overriding functionality inherited from a parent class.
`virtual` should be used on the first class in the hierarchy that defines the method.

SteffenMath replied:
Okay I'm going to fix that.

pcarruscag commented on SU2_CFD/include/drivers/CSinglezoneDriver.hpp:
```diff
@@ -98,6 +98,12 @@ class CSinglezoneDriver : public CDriver {
+   * \brief gets Convergence on physical time scale
+   * \param none
+   */
+  virtual bool GetTimeConvergence() const;
```
What are the implications for multizone problems?

SteffenMath replied:
At the moment, the convergence tools only work for single-zone problems, however it should be easy to extend it. I will speak about it with @talbring.

pcarruscag replied:
Ok. If you want to merge the PR before that is done please add an error message or warning somewhere if someone tries to use the feature in multizone problems.

SteffenMath replied:
I've added an error message for this case.

pcarruscag commented on TestCases/disc_adj_rans/cylinder/cylinder_Windowing.cfg:
```diff
@@ -0,0 +1,416 @@
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
```
As a general note on examples, I would try to clean the config as much as possible to really put the focus on the new options you added, for example if MG_LEVEL=0 all other MG options don't do anything, if you use the Roe scheme the JST coefficients are not important and so on.

SteffenMath replied:
I have removed default options, that are unnecessary for a meaningful test case.

TobiKattmann commented:
I'll read through the PR over the weekend, so I would appreciate this staying open until Monday.

SteffenMath commented on Common/include/config_structure.hpp:
```diff
@@ -5425,6 +5434,19 @@ class CConfig {
 
+  /*!
+   * \brief Get the Starting Iteration for the windowing approach
+   *        in Sensitivity Analysis for LCOs.
```
LCO means Limit Cycle Oscillator, I admit, that this is not a common term. 
I changed the description to something more self-explainatory.

SteffenMath commented on SU2_CFD/include/drivers/CDiscAdjSinglezoneDriver.hpp:
```diff
@@ -127,4 +127,10 @@ class CDiscAdjSinglezoneDriver : public CSinglezoneDriver {
+  /*!
+   * \brief gets Convergence on physical time scale, (deactivated in adjoint case)
+   * \param none
+   */
```
Done

SteffenMath commented on SU2_CFD/include/output/COutput.hpp:
```diff
@@ -150,8 +150,8 @@ class COutput {
-  map<string, Signal_Processing::RunningAverage> runningAverages;
-
+  map<string, CWindowedAverage> windowedTimeAverages;
+  
```
Removed

SteffenMath commented on SU2_CFD/include/output/COutput.hpp:
```diff
@@ -413,6 +425,14 @@ class COutput {
+     * \param config, Iteration
+     */
+  bool MonitorTimeConvergence(CConfig *config, unsigned long Iteration);
+  bool GetTimeConvergence()const {return TimeConvergence;} /*! \brief Indicates, if the time loop is converged. COnvergence criterion: Windowed time average */
```
Done

SteffenMath commented on SU2_CFD/include/output/tools/CWindowingTools.hpp:
```diff
@@ -0,0 +1,67 @@
+  su2double val;
+  unsigned long count;
+  std::vector<su2double> values;
+
```
I added commentary to all member functions and variables.

SteffenMath commented on SU2_CFD/include/output/tools/CWindowingTools.hpp:
```diff
@@ -0,0 +1,67 @@
+
+  su2double Update(su2double valIn); //Legacy
+  su2double Get() const;
+  unsigned long Count() const; //Legacy
```
I removed the (unused) legacy code, so this is no longer an issue :)

SteffenMath commented on SU2_CFD/src/drivers/CDiscAdjMultizoneDriver.cpp:
```diff
@@ -775,20 +775,20 @@ void CDiscAdjMultizoneDriver::SetObjFunction(unsigned short kind_recording) {
+  if (time_stepping){
     if (TimeIter < IterAvg_Obj){
-      seeding = 1.0/((su2double)IterAvg_Obj);
+      seeding = windowEvaluator.GetWndWeight(config_container[ZONE_0]->GetKindWindow(),TimeIter, IterAvg_Obj-1)/ (static_cast<su2double>(IterAvg_Obj));
```
Ok done

SteffenMath commented on SU2_CFD/src/drivers/CDiscAdjSinglezoneDriver.cpp:
```diff
@@ -270,9 +271,11 @@ void CDiscAdjSinglezoneDriver::SetAdj_ObjFunction(){
   if (time_stepping){
     if (TimeIter < IterAvg_Obj){
-      seeding = 1.0/((su2double)IterAvg_Obj);
+      seeding = windowEvaluator.GetWndWeight(config->GetKindWindow(),TimeIter, IterAvg_Obj-1)/ (static_cast<su2double>(IterAvg_Obj));
```
Done as well :)

SteffenMath commented on SU2_CFD/src/drivers/CSinglezoneDriver.cpp:
```diff
@@ -308,6 +311,10 @@ bool CSinglezoneDriver::Monitor(unsigned long TimeIter){
 
+bool CSinglezoneDriver::GetTimeConvergence() const{
+  return output_container[ZONE_0]->GetTimeConvergence();
+}
```
I've changed the function definition to an inline definition, as @pcarruscag has suggested.

SteffenMath commented on SU2_CFD/src/meson.build:
```diff
@@ -63,7 +63,8 @@ su2_cfd_src += files(['output/CAdjElasticityOutput.cpp',
                       'output/filewriter/CSU2BinaryFileWriter.cpp',
-                      'output/filewriter/CSU2MeshFileWriter.cpp'])
+                      'output/filewriter/CSU2MeshFileWriter.cpp',
+					  'output/tools/CWindowingTools.cpp'])
```
I used spaces now. Hope this resolves the issue.

SteffenMath commented on SU2_CFD/src/output/tools/CWindowingTools.cpp:
```diff
@@ -0,0 +1,142 @@
+  su2double endTime = static_cast<su2double>(endTimeIdx);
+  su2double tau = currTime/endTime;
+  return 1.0-cos(2*PI_NUMBER*tau);
+}
```
Okay done.

SteffenMath commented on SU2_CFD/src/output/tools/CWindowingTools.cpp:
```diff
@@ -0,0 +1,142 @@
+
+su2double CWindowedAverage::Get() const{
+  return val;
+}
```
Okay done.

SteffenMath commented on SU2_CFD/src/output/tools/CWindowingTools.cpp:
```diff
@@ -0,0 +1,142 @@
+  return val;
+}
+
+unsigned long CWindowedAverage::Count() const{
```
This function has been removed, since it is unused.  This resolves this issue :)

TobiKattmann commented on SU2_CFD/include/drivers/CSinglezoneDriver.hpp:
```diff
@@ -98,6 +98,15 @@ class CSinglezoneDriver : public CDriver {
+   * \brief gets Convergence on physical time scale
+   * \param none
+   */
+  inline virtual bool GetTimeConvergence() const{
```
Comment cleanup

SteffenMath replied:
Done

TobiKattmann commented on SU2_CFD/include/output/tools/CWindowingTools.hpp:
```diff
@@ -0,0 +1,120 @@
+  * \param endTimeIdx - Number of time steps to average over
+  * \return Value of the window-weight-function at time CurTimeIdx with end-time endTimeIdx
+  */
+  su2double HannWindow(unsigned long i, unsigned long endTimeIdx) const;
```
As far as I understand you are using unsigned long `i` `CurTimeIdx` and `currentIter` in this file which all have the same meaning : `Current time iteration of the solver`
For consistency it would be nice to choose one for your implementation, preferably `CurTimeIdx` or something new `CurrTimeIter`

SteffenMath replied:
I replaced the variable-names in questions with `curTimeIter`.

TobiKattmann commented on SU2_CFD/include/output/tools/CWindowingTools.hpp:
```diff
@@ -0,0 +1,120 @@
+   * \param endTimeIdx - Number of time steps to average over
+   * \return Value of the window-weight-function at time CurTimeIdx with time-frame endTimeIdx
+   */
+  su2double GetWndWeight(WINDOW_FUNCTION windowId, unsigned long CurTimeIdx, unsigned long endTimeIdx) const;
```
here (in the \brief) it is probably `windowId` instead of `fctIdx`?

SteffenMath replied:
You are right.
